lib/matplotlib/axes/_axes.py

```diff
@@ -4382,6 +4406,25 @@ def do_patch(xs, ys, **kwargs):
                 flier_y = stats['fliers']
                 fliers.append(do_plot(flier_x, flier_y, **flier_kw))
 
+        # Set legend labels
+        if label:
+            box_or_med = boxes if showbox else medians
+            for index, element in enumerate(box_or_med):
+                if cbook.is_scalar_or_string(label):
+                    # If there are multiple boxes, the label
+                    # should only be applied to the first box.
+                    if len(box_or_med) > len(label):
+                        element.set_label(label)
+                        break
+                    else:
+                        element.set_label(label)
+                else:  # label is a sequence
+                    if len(box_or_med) == len(label):
+                        element.set_label(label[index])
+                    else:
+                        raise ValueError("There must be an equal number of legend"
+                                         " labels and boxplots.")
```

Comment:
This is not correct. For example change your test below `bp4 = axs[3].boxplot(new_data, label='A')` to a label with more than one char and it will fail. Also you can make this more understanable (and fix the behavior) by moving all the logic out of the loop:

```suggestion
            if cbook.is_scalar_or_string(label):
                # assign the label only to the first box
                box_or_med[0].set_label(label)
            else:  # label is a sequence 
                if len(box_or_med) != len(label):
                    raise ValueError("There must be an equal number of legend"
                                   " labels and boxplots.")
                for artist, lbl in zip(box_or_med, label):
                    artist.set_label(lbl)
```